simphony/testing/abc_check_lattice.py

```diff
@@ -0,0 +1,229 @@
+        """ Create and return a lattice.
+        """
+
+    @abc.abstractmethod
+    def supported_cuba(self):
+        """ Return a list of CUBA keys to use for restricted containers.
+
+        """
+
+    def test_iter_nodes(self):
+        container = self.container
+
+        # number of nodes
+        number_of_nodes = sum(1 for node in container.iter_nodes())
+        self.assertEqual(number_of_nodes, numpy.prod(self.size))
+
+        # data
+        for node in container.iter_nodes():
+            self.assertEqual(node.data, DataContainer())
+
+        # indexes
+        x, y, z = numpy.meshgrid(
+            range(self.size[0]), range(self.size[1]), range(self.size[2]))
+        expected = set(zip(x.flat, y.flat, z.flat))
+        indexes = {node.index for node in container.iter_nodes()}
+        self.assertEqual(indexes, expected)
+
+    def test_iter_nodes_subset(self):
+        container = self.container
+
+        x, y, z = numpy.meshgrid(
+            range(2, self.size[0]),
+            range(self.size[1]-4),
+            range(3, self.size[2], 2))
+        expected = set(zip(x.flat, y.flat, z.flat))
+
+        # data
+        for node in container.iter_nodes(expected):
+            self.assertEqual(node.data, DataContainer())
+
+        # indexes
+        indexes = {node.index for node in container.iter_nodes(expected)}
+        self.assertEqual(indexes, expected)
+
+    def test_get_node(self):
+        container = self.container
+
+        index = 2, 3, 4
+        node = container.get_node(index)
+        expected = LatticeNode(index)
+        self.assertEqual(node, expected)
+
+        # check that mutating the node does not change internal info
+        node.data = create_data_container()
+        self.assertNotEqual(container.get_node(index), node)
+
+    def test_get_node_with_invalid_index(self):
+        container = self.container
+
+        index = 2, 300, 4
+        with self.assertRaises(IndexError):
+            container.get_node(index)
+
+        index = 2, 3, -4
+        with self.assertRaises(IndexError):
+            container.get_node(index)
+
+    def test_update_node(self):
+        container = self.container
+
+        index = 2, 3, 4
+        node = container.get_node(index)
+        node.data = create_data_container(restrict=self.supported_cuba())
+        container.update_node(node)
+
+        new_node = container.get_node(index)
+        self.assertEqual(new_node, node)
+        # Check that `new_node` is not the same instance as `node`
+        self.assertIsNot(new_node, node)
+
+    def test_update_node_with_extra_keywords(self):
+        container = self.container
+
+        index = 2, 3, 4
+        node = container.get_node(index)
+        # Update with full DataContainer.
+        node.data = create_data_container()
+        container.update_node(node)
+
+        new_node = container.get_node(index)
+        # We expect only the supported CUBA to be stored.
```

Comment:
What if we have a CustomRecord table with not all LBM supported CUBA keys defined? Does this test fail then?